### PR TITLE
Sphinx docs, Auto-build to GitHub pages, and Functional changes

### PR DESCRIPTION
**Pull request by ItsSimko**
# Sphinx docs, Auto-build to GitHub pages, and misc changes

### Description
Created sphinx documentation to auto-deploy to GitHub actions. Also number of functional changes

### Changes Made
- Created sphinx documentation
- Added GitHub workflow to auto-deploy to GitHub pages
- Changed docker containers to run as non-root
- Gave Caddy and Aphrodite each their own Dockerfile that still pulls images but allows more of our customizations
- Moved the service off of port 443 onto 3334
- The LLM-Container is no longer accessible via HTTP only HTTPS via caddy on port 3334
- Added environmental var to change the port to access the server on LLM_CONTAINER_PORT

## Summary by Sourcery

Implement Sphinx documentation with auto-deployment to GitHub Pages, enhance Docker container security by running as non-root, customize Dockerfiles for Caddy and Aphrodite, and update service port configurations for improved security and flexibility.

New Features:
- Introduce Sphinx documentation for the project, enabling structured and comprehensive documentation generation.

Enhancements:
- Modify Docker containers to run as non-root users, enhancing security.
- Assign separate Dockerfiles to Caddy and Aphrodite, allowing for more customization while still pulling base images.
- Change the service port from 443 to 3334, and ensure the LLM-Container is accessible only via HTTPS through Caddy on the new port.
- Add an environment variable to allow changing the server access port via LLM_CONTAINER_PORT.

Build:
- Add a GitHub workflow to automatically deploy the Sphinx documentation to GitHub Pages.

Documentation:
- Create a Sphinx configuration file and supporting scripts (Makefile and make.bat) to facilitate building and managing the documentation.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Configuration file for the Sphinx documentation builder.
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+# -- Project information -----------------------------------------------------
+# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information
+
+project = 'Local-LLM-Container'
+copyright = '2024, ClinicianFOCUS'
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply docs/conf.py:10]: This is not controlled by us but by sphinx.

File: docs/index.rst [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+To stop the containers:
+
+.. code-block:: bash
+
+   docker-compose down
+
+This will stop and remove the containers, but the built images and mounted volumes will persist.
+
+
+Additonal Notes
```
[the remark this answers is not in the document]
Comment (yingbull) [reply docs/index.rst:108]: Why not an offer for a commit to fix, Sourcery?

File: docs/index.rst [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+Follow these steps to set up and run the services using Docker Compose:
+
+1. **Clone the repository**:
+
+   .. code-block:: bash
+
+      git clone https://github.com/ClinicianFOCUS/local-llm-container.git
+      cd local-llm-container
+
+2. **Ensure Docker and Docker Compose are installed**. You can install Docker by following the instructions on the [official Docker website](https://docs.docker.com/get-docker/).
```
[the remark this answers is not in the document]
Comment (yingbull) [reply docs/index.rst:42]: Just make sure we have the markdown plugin with the builder from Sphynx I think, @ItsSimko

File: docker-compose.yml [hunk 4]
```diff
@@ -1,27 +1,57 @@
...
           devices:
             - driver: nvidia
               count: all
               capabilities:
                 - gpu
+    build:
+      # Build context and Dockerfile for the LLM container
+      context: .
+      dockerfile: Dockerfile.llm-cont
+    user: "1000:1000" # Set the user to '1000:1000'
```
[the remark this answers is not in the document]
Comment (yingbull) [reply docker-compose.yml:19]: This was intentional in this case and tested to work.

File: docker-compose.yml [hunk 5]
```diff
@@ -1,27 +1,57 @@
...
+    # Ensure the container runs with the specified user and group ID
     volumes:
-      - ~/.cache/huggingface:/root/.cache/huggingface
+      # Mount the local models directory to the container's /models directory
       - ./models:/models
-    ports:
-      - 2242:2242
+    networks:
+      # Attach the container to the llm-network
+      - llm-network
```
[the remark this answers is not in the document]
Comment (yingbull) [reply docker-compose.yml:26]: Our setup is correct.

File: docs/index.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply docs/index.rst:108]: You do it for other ones you suggest, Soucery - give me an offer to commit the suggestion.

**Review by yingbull — APPROVED** [review]

**Pull request merged**